File: legacy/src/main/java/org/yankov/mso/datamodel/DurationConverter.java
package org.yankov.mso.datamodel;

import javax.persistence.AttributeConverter;
import javax.persistence.Converter;
import java.time.Duration;

@Converter
public class DurationConverter implements AttributeConverter<Duration, String> {

    private static final String SEPARATOR = ":";
    private static final String FORMAT_TEMPLATE = "%02d" + SEPARATOR + "%02d" + SEPARATOR + "%02d";

    private static final long HOURS_IN_DAY = 24;
    private static final long MINUTES_IN_HOUR = 60;
    private static final long SECONDS_IN_MINUTE = 60;
    private static final long SECONDS_IN_HOUR = 60 * SECONDS_IN_MINUTE;
    private static final long MILLIS_IN_SECOND = 1000;

    @Override
    public String convertToDatabaseColumn(Duration duration) {
        return duration != null ? String
            .format(FORMAT_TEMPLATE, toHoursPart(duration), toMinutesPart(duration), toSecondsPart(duration)) : "";
    }

    @Override
    public Duration convertToEntityAttribute(String s) {
        if (s == null || s.isEmpty()) {
            return null;
        }

        String[] elements = s.split(SEPARATOR);

        int hoursPart = Integer.parseInt(elements[0]);
        int minutesPart = Integer.parseInt(elements[1]);
        int secondsPart = Integer.parseInt(elements[2]);

        long seconds = hoursPart * SECONDS_IN_HOUR + minutesPart * SECONDS_IN_MINUTE + secondsPart;

        return Duration.ofSeconds(seconds);
    }

    public static long toHoursPart(Duration duration) {
        return duration.toHours() % HOURS_IN_DAY;
    }

    public static long toMinutesPart(Duration duration) {
        return duration.toMinutes() % MINUTES_IN_HOUR;
    }

    public static long toSecondsPart(Duration duration) {
        return (duration.toMillis() / MILLIS_IN_SECOND) % SECONDS_IN_MINUTE;
    }

}

File: src/main/scala/org/yankov/mso/application/converters/DurationConverter.scala
package org.yankov.mso.application.converters

import java.time.Duration

object DurationConverter {
  private val separator = ':'
  private val hhMMss = "%02d" + separator + "%02d" + separator + "%02d"
  private val hMMss = "%d" + separator + "%02d" + separator + "%02d"
  private val minutesInHour = 60
  private val secondsInMinute = 60
  private val secondsInHour = 60 * secondsInMinute
  private val millisInSecond = 1000

  def toHourMinSecString(duration: Duration, trimLeadingZeros: Boolean = false): String = {
    val f = if (trimLeadingZeros) hMMss else hhMMss
    val str = f.format(duration.toHours, toMinutesPart(duration), toSecondsPart(duration))
    if (trimLeadingZeros && duration.toHours == 0) str.substring(str.indexOf(separator) + 1)
    else str
  }

  def fromString(s: String): Duration = {
    if (s.isEmpty) Duration.ZERO
    else {
      val elements = s.split(separator.toString)

      val hoursPart = Integer.parseInt(elements(0))
      val minutesPart = Integer.parseInt(elements(1))
      val secondsPart = Integer.parseInt(elements(2))

      val seconds = hoursPart * secondsInHour + minutesPart * secondsInMinute + secondsPart

      Duration.ofSeconds(seconds)
    }
  }

  private def toMinutesPart(duration: Duration): Long = duration.toMinutes % minutesInHour

  private def toSecondsPart(duration: Duration): Long = (duration.toMillis / millisInSecond) % secondsInMinute
}
